python/ray/autoscaler/gcp/node_provider.py

```diff
@@ -0,0 +1,194 @@
+        # Note: All the operations use 'name' as the unique instance identifier
+        self.cached_nodes = {i['name']: i for i in instances}
+
+        return [i['name'] for i in instances]
+
+    def is_running(self, node_id):
+        node = self._node(node_id)
+        return node['status'] == 'RUNNING'
+
+    def is_terminated(self, node_id):
+        node = self._node(node_id)
+        return node['status'] in TERMINATED_STATES
+
+    def node_tags(self, node_id):
+        node = self._node(node_id)
+        labels = node.get('labels', {})
+        return labels
+
+    def set_node_tags(self, node_id, tags):
+        labels = tags
+        project_id = self.provider_config['project_id']
+        availability_zone = self.provider_config['availability_zone']
+
+        node = self._node(node_id)
+        operation = self.compute.instances().setLabels(
+            project=project_id,
+            zone=availability_zone,
+            instance=node_id,
+            body={
+                'labels': dict(node['labels'], **labels),
+                'labelFingerprint': node['labelFingerprint']
+            }
+        ).execute()
+
+        result = wait_for_compute_zone_operation(
+            project_id, operation, availability_zone)
+
+        return result
+
+    def external_ip(self, node_id):
+        if node_id in self.external_ip_cache:
+            return self.external_ip_cache[node_id]
+        node = self._node(node_id)
+        # TODO: Isn't there better and more reliable way to do this?
+        ip = (node
+              .get('networkInterfaces', [{}])[0]
+              .get('accessConfigs', [{}])[0]
+              .get('natIP', None))
+        if ip:
+            self.external_ip_cache[node_id] = ip
+        return ip
+
+    def internal_ip(self, node_id):
+        if node_id in self.internal_ip_cache:
+            return self.internal_ip_cache[node_id]
+        node = self._node(node_id)
+        ip = node.get('networkInterfaces', [{}])[0].get('networkIP')
+        if ip:
+            self.internal_ip_cache[node_id] = ip
+        return ip
+
+
+    def create_node(self, base_config, tags, count):
+        # NOTE: gcp uses 'labels' instead of aws 'tags'
+        # https://cloud.google.com/compute/docs/instances/create-start-instance#startinginstancwithimage
+        labels = tags
+        project_id = self.provider_config['project_id']
+        availability_zone = self.provider_config['availability_zone']
+
+        config = base_config.copy()
+        config['name'] = labels[self.tag_keys['node-name']]
+        config['machineType'] = (
+            'zones/{zone}/machineTypes/{machine_type}'
+            ''.format(zone=availability_zone,
+                      machine_type=base_config['machineType']))
+
+        config['labels'] = dict(
+            config.get('labels', {}),
+            **labels,
+            **{self.tag_keys['cluster-name']: self.cluster_name})
+
+        if count != 1: raise NotImplementedError(count)
```

Comment:
It's also ok to just create 1 (and change this to a warning) -- AWS never guarantees if you try to create N nodes you actually that many created. This possible discrepancy is already handled in the autoscaling loop.